### ISSUE TITLE
Parameters for Media Types

### ISSUE THREAD

**Issue by cjaccino** [OAI/OpenAPI-Specification#2342]
### Parameters for Media Types

Media Types can support parameters.  Yet, to work with them in OpenAPI, one must define them against each media type.  This has the unfortunate side effect of discouraging use of media type parameters by making them clumsy to manage within the API spec.  

```
content:
  text/plain; charset=utf-8:
    ...
  text/plain; charset=iso-8859-1:
    ...
```

Without a way to express the parameter, it has little or no semantic value for OpenAPI.  Minor variations look like completely different media types.  There is little room for negotiation over content parameters unless the API author takes the time to spell out each supported parameter as its own content type.

Media Type parameters could be useful in some obvious ways:
  - versioning of the media type
  - specifying a 'profile' for the media type as described in RFC6906
  - negotiating support for embedded resources (i.e. HAL (https://tools.ietf.org/html/draft-kelly-json-hal-08)) , hypermedia style, or level of completeness of a representation.

Private and vendor media types might better manage breaking changes by supporting multiple versions of representation without ever changing endpoints and methods.  Perhaps there could be benefits to QA and regression test tooling, being able to validate support for the diversity of representation styles in the API contract.  This could be had without having to iterate every parameter combination as a separate media type.

### Examples
**Versioning**
```
Accept: application/vnd.example-schema+json; version=1.0.4
```

```
content:
  application/vnd.example-schema+json:
    parameters:
      version:
        type: string
        required: false
        description: |
          The version of the representation.
        default: 1.1.0
        enum:
          - 1.0.3
          - 1.0.4
          - 1.1.0
    schema:
    	$ref: "https://schemas.example.com/example"
```

**Profile identification**
```
Accept: application/vnd.example-schema+json; profile=https://schemas.example.com/example/profiles/bob
```

```
content:
  application/vnd.example-schema+json:
    parameters:
      profile:
        type: string
        required: false
        description: |
          The profile (rfc6906) for the representation.
        default: "https://schemas.example.com/example/profiles/alice"
        enum:
          - "https://schemas.example.com/example/profiles/alice"
          - "https://schemas.example.com/example/profiles/bob"
    schema:
    	$ref: "https://schemas.example.com/example"
```

**Hypermedia style selection**
```
Accept: application/vnd.example-schema+json; hypermedia=HAL
```
```
content:  
  application/vnd.example-schema+json:
    parameters:
      hypermedia:
        type: string
        required: false
        description: |
          Hypermedia Style.  Indicates whether links and other supplemental
          resource detail are provided in the representation.
        default: JSON_HYPERSCHEMA
        enum:
          hypermedia:
            - NONE
            - HAL
            - JSON_HYPERSCHEMA
    schema:
    	$ref: "https://schemas.example.com/example"
```

**Comment by philsturgeon** [comment]
I'm right there with you on parameters and your examples for versioning and profile identification,  but I think the hypermedia selection stuff is non-standard and we don't want to add anything non-standard to OpenAPI thats going to live in the users actual API.

If we stick the a general mechanism for adding parameters to a media type, I think this could be a great idea. charset, version, etc are all good things to be able to describe as possible in an API.

**Comment by cjaccino** [comment]
Thanks Phil.
You're right, the hypermedia example is flawed.  I'm happy to remove it from the issue or provide other examples if it's helpful.

**Comment by MikeRalphson** [comment]
For consistency, if someone does take this to a PR, fields like `type` and `enum` should be wrapped within a standard `schema` field of type `schemaObject`.

**Comment by cjaccino** [comment]
Right. All of the examples need work.

The operation ```parameters``` field may be the best thing to compare for consistency.  

Guessing at the changes that might be needed:
 - A new ```Media Type Parameter Object```.  Or possibly a subset of the existing Parameters Object?
 - Add ```parameters``` to the Media Type Object whose type would be ```Map[Media Type Parameter Object | Reference Object]```.
 - JSON Schema updates.  Happy to propose these; would be looking for advice on norms for contributing to the project beyond issue contributions.  

### Revised Examples
**Versioning**
```http
Accept: application/vnd.example-schema+json; version=1.0.4
```

```yaml
content:
  application/vnd.example-schema+json:
    parameters:
      - name: version
         description:  Media type version for the representation.
         required: false
         deprecated: false
         schema:
           type: string
           default: 1.1.0
           enum:
             - 1.0.3
             - 1.0.4
             - 1.1.0
         examples:
           priorVersion:
             summary:  The most recent prior version.
             value: 1.0.4
           currentVersion:
             summary:  The current version.
             value: 1.1.0
```
or
```yaml
content:
  application/vnd.example-schema+json:
    parameters:
      - $ref: "https://schemas.example.com/example/MediaTypeVersion"
```

**Profile identification**
```http
Accept: application/vnd.example-schema+json; profile=https://schemas.example.com/example/profiles/bob
```

```yaml
content:
  application/vnd.example-schema+json:
    parameters:
      - name: profile
       description: The profile (rfc6906) for the media type
       required: false
       deprecated: false
       schema: 
         type: string
         default: "https://schemas.example.com/example/profiles/alice"
         enum:
           - "https://schemas.example.com/example/profiles/alice"
           - "https://schemas.example.com/example/profiles/bob"
        examples:
          BobProfile:
            summary:  The classic Bob profile we all know and love.
            value: "https://schemas.example.com/example/profiles/bob"
          AliceProfile:
            summary:  The new and improved.
            value: "https://schemas.example.com/example/profiles/alice"
```
_Use of ```value```, not ```externalValue```, as the value is intended to be the URL string, not the profile schema to which the URL refers.  Interested if there are disagreements._

or
```yaml
content:
  application/vnd.example-schema+json:
    parameters:
      - $ref: "https://schemas.example.com/example"
```

**Comment by darrelmiller** [comment]
My initial reaction to Phil's tweet, was "OMG No!" .  The opaque media type identifier was by design.  However,  having read this issue and seeing the examples, I am warming up to the idea.
The parameters array is nice for enumerating the possible parameters and the providing the syntax.  I have to admit I can never remember the syntax of `application/json;odata.metadata=full`.

However, once we get into the territory of profile and version parameters, we get to a point where we almost need to be able to have different schemas and examples based on the parameter values.  What's the point of documenting profiles if you can't also document the shape that the profile identifies.   I'm not sure how you could add that cleanly.

**Comment by cjaccino** [comment]
Looking at non-structured media types may put parameters in a different light.

Drawing from [the media type registration for PCMU audio](https://tools.ietf.org/html/rfc4856#section-2.1.18).  I'm ignoring the fact that the registration assumed PCMU's use in RTP;  the same media type might be used to interact with stored audio prompts, songs, etc. via HTTP.

```http
GET /songs/happy-birthday
Accept: audio/PCMU;rate=44100;channels=2;q=0.9, audio/PCMU;rate=8000;q=0.8, audio/*;q=0.5
```

```yaml
content:
  audio/PCMU:
    parameters:
      - name: rate
        description: The bitrate for the audio
        required: true
        schema: 
           type: string
           default: "8000"
           enum:
             - "6000"
             - "8000"
             - "11025"
             - "16000"
             - "22050"
             - "32000"
             - "44100"
             - "48000"
             - "64000"
             - "88200"
             - "96000"
             - "176400"
             - "192000"
      - name: channels
        description: The number of audio channels
        required: false
        schema:
          type: integer
          default: 1
          enum:
            - 1
            - 2
```
### Benefit to API Author and Client Developer
To be realistic, I took the options in the bitrate enum come from Adobe Audition. With thirteen bitrate options and a mono/stereo option, OpenAPI 3.0 requires twenty-six independent media types to express the options. Not only is it inefficient, but it implicitly and wrongly suggests that the multiple media types are different. They are not.

On the client side, the cost of reducing the options into meaningful handling is higher than it ought to be. In this example, not being able to express this aspect of HTTP forces the author to either take on inefficiency or solve the same problem by other means. Query parameters are probably the next best option, but if the media type authors intended for clients to interact with the resource and representations using parameters, why force the API author to develop an alternative?

### Addressing Examples
The examples question is interesting. Using the plural examples: object, one has the opportunity to name each example. There isn't explicit mapping (yet proposed). Authors have always had to question how far to go with examples. JSON already provides so much flexibility that documenting all permutations may be unrealistic. I'm not sure mapping the examples to parameters is a problem that must be solved; maybe OpenAPI already has what it needs for examples.

### The Value of Versioning
Versioning and profiles are just examples. It may not even be good practice to version some media types; I've seen commentary in this direction. But having versioning could enable organizations offering SLA-supported API products to gracefully revise their APIs' private media types and schemas without changing the API structurally. In this case, media type parameters would enable client developers to step through schema updates gracefully, selecting the supported earlier representation while being informed of updates. 

I hope we can improve OpenAPI's semantic support for Content Negotiation, one of HTTP's most powerful features. Best practices for defining media types will continue to evolve, and our examples are probably just scratching the surface.  Looking back to Roy Fielding's statement on media types, I believe this is consistent with his intent.

**Comment by karenetheridge** [comment]
I ran into this problem ("there are multiple media types but they're really all the same") when implementing body deserialization for text-style media types with different charsets:

- `Content-Type: text/plain; charset=UTF-8` and `Content-Type: text/plain; charset=iso-2022-jp-1` are both just text
- ultimately I want to get a unicode text string that I can apply against my json schema to get validation results
- therefore I MUST parse that charset parameter while deserializing the HTTP request
- but the JSON Schema for the text is going to be the same for each charset; the client doesn't really give a damn how the string was encoded, as it wants the decoded version
- so I need to create duplicate media type entries in the OAD (which I can't even do with `$ref`s since there is no facility in the spec for references to a media-type object), AND I need to add special code into the HTTP message parser to recognize and make use of the charset parameter anyway

My choice here was to opt to stripping everything after the `;` in the media type, so all variations map to the same media type entry of "text/plain", and then deal with the charset decoding as I would need to do anyway.

I like the invention here of a "parameter" property in the Media-Type Object, which is especially useful for client generation tools to give a guide as to what sort of values the server is likely to accept (e.g. audio bitrates, compression schemes, etc), as well as allow for a server to filter out and reject unsupported parameters before the core application ever sees it.

Can we move forward with this proposal and get on with the bits we enjoy most: debating whether the `parameters` objects should be an array under the Media-Type Object or objects whose keys are their names, and how `$ref`s should work and whether to allow a new object type under `/components` and what it should be called?

**Comment by karenetheridge** [comment]
..Adding a note here that the requirement for charset decoding is not mentioned in the specification and it really should be; there may be other media-type parameters that should also receive treatment that I can't think of right now.

**Comment by handrews** [comment]
@karenetheridge 

> Adding a note here that the requirement for charset decoding is not mentioned in the specification and it really should be

I'm not sure I understand what requirement you mean?

> which I can't even do with $refs since there is no facility in the spec for references to a media-type object

PR #4728 

> therefore I MUST parse that charset parameter while deserializing the HTTP request

But you don't really have to get it from the OAD, it would have to be in the `Content-Type` header anyway.

> Can we move forward with this proposal

I certainly won't stop any discussion, but I'm a bit consumed by trying to get 3.2 out the door at the moment.

**Comment by handrews** [comment]
Also 

>  the bits we enjoy most: debating whether the parameters objects should be an array under the Media-Type Object or objects whose keys are their names

🤣 🤣 🤣

**Comment by karenetheridge** [comment]
> > Adding a note here that the requirement for charset decoding is not mentioned in the specification and it really should be

> I'm not sure I understand what requirement you mean?

I suppose it's implicit that (edit: not) only the base media-type in the Content-Type header is used for decoding, but the charset as well.  Certainly any webserver framework would be taking care of this for the user, but one might be doing it manually.

**Comment by ericmorand** [comment]
Would this change cover the case where we need different header and cookie parameters, based on the media type?

For example, let's consider a RESTful API where a resource is identified by an URI, and the representation of the resource is controlled by the media type. Depending on the requested representation, the parameters _may_ be different. A User resource may require:

* a cookie parameter when the representation is `text/html`
* a header parameter (e.g. `Authorisation: Bearer <token>`)  when the representation is `application/json`

Currently, this does not seem to be possible. The following document is not valid:

```json
{
  "openapi": "3.1.1",
  "info": {
    "title": "OpenID Connect Playground",
    "version": "1.0.0"
  },
  "paths": {
    "/user": {
      "get": {
        "summary": "GetUser",
        "description": "Return the authenticated user.",
        "responses": {
          "200": {
            "description": "The authenticated user",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/user",
                  "description": "The JSON representation of the authenticated user"
                },
                "parameters": [
                  {
                    "name": "Authorization",
                    "in": "header",
                    "schema": {
                      "type": "string"
                    }
                  }
                ]
              },
              "text/html": {
                "schema": {
                  "type": "string",
                  "description": "The HTML representation of the authenticated user"
                },
                "parameters": [
                  {
                    "name": "access_token",
                    "in": "cookie",
                    "schema": {
                      "type": "string"
                    }
                  },
                  {
                    "name": "refresh_token",
                    "in": "cookie",
                    "schema": {
                      "type": "string"
                    }
                  }
                ]
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "user": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "email": {
            "type": "string"
          }
        },
        "required": [
          "name",
          "email"
        ],
        "additionalProperties": false
      }
    }
  }
}
```

The current specification enforces to share the same parameters accross the representations.

Would the current proposal sole this issue too?

**Comment by handrews** [comment]
@ericmorand no, what you're talking about is interdependencies among parameters, which is not supported at all up through 3.1, and is only supported for the query string in 3.2 with the new `in: querystring` location.  There is an issue filed for this somewhere (probably moved over to the OAI/sig-moonwalk repository as a Discussion there, as that's currently where we work out how to approach bigger changes).

This issue is about media type parameters, so `charset` in `application/xml; charset=utf-8` or `schema` in `application/schema-instance+json; schema=https://example.com/foo/schema`.  There are some more elaborate examples in earlier comments on this issue.

**Comment by karenetheridge** [comment]
This feature request is at the top of my list for 3.3, as I have noticed an ambiguity in the spec where I had to make a decision for my implementation:

I frequently deal with requests+responses which use `Content-Type: application/json; charset=UTF-8` -- but the amount of whitespace after the semicolon can be variable, and there could be other parameters as well (that I may or may not care about).  So it seemed to me that using that full string to look up the media type (under the `content` field) is incorrect, so I strip off everything after the first semicolon in the Content-Type header before performing the matching.

Further, for `text/*` media types only (there is an RFC somewhere that says that charset should only apply to text types; exact reference left as an exercise for the reader), I look for a `charset` parameter and apply that to the content before passing it along to the JSON Schema for validation. Again, if I was to follow the spec strictly here, I'd have to enumerate every possible charset under `content` and provide a different media type decoder for each one (each of which only varied by the charset in question), which seems wrong. Applying the parameters afterwards (after applying the media-type decoding) seems much more natural than defining individual media types that include all the possible parameter variations.

edit: it looks like I mostly just repeated things I've already said earlier on this issue. sorry about that :)

**Comment by handrews** [comment]
@karenetheridge 

TL;DR: You have lots of reasonable concerns but I have a differing view of some specific points, and maybe some examples to help show the distinctions.

 > So it seemed to me that using that full string to look up the media type (under the content field) is incorrect, so I strip off everything after the first semicolon in the Content-Type header before performing the matching.

The more correct thing to do would be to parse the parameters and treat an entry with paramters as more specific than one without them.  We very specifically note that we are comparing media ranges, which requires parsing them to determine that `application/json` is more specific than `application/*`. And, perhaps, `application/foo+json` is more specific than `application/json`, although I don't think anyone has asked about that.  So it's never been just a string comparison.

> Again, if I was to follow the spec strictly here, I'd have to enumerate every possible charset under content and provide a different media type decoder for each one

I don't believe that's what the spec says (although see below as `charset` has other aspects that I think are leading you to this conclusion).  The spec says more specific takes precedence over more general, which implies that more general media ranges match if no more specific one is available.  So if only `text/plain` is specified and you get `text/plain; charset=utf-8`, you match `text/plain`.  No ambiguity there at all.  If you have:

```yaml
content:
  text/plain: {}
  text/plain; charset=ebcdic: {}
```

and get `text/plain; charset=utf-8`, then you match `text/plain`.

-----

None of this addresses your valid concern (that I recall from somewhere) about what parameters are _relevant_, though, or how to manage a confusing combination of parameters.

With `charset`, there's really nothing to do in terms of "provid[ing] a different media type decoder."  The `charset` parameter would only be used to match the correct Media Type Object, but you would handle parsing the body the same regardless of what keys are specified under `content`.  Because nothing in the Media Type Object actually cares about the character set (except examples in `serializedExample` or `externalExample`, but those are now well-defined).

So I would think you'd only have one `text/plain` decoder, it's just that you might match different Media Type Objects depending on `charset`... and then use the same decoder either way.  I think.  LMK if I'm way off base here, it's entirely possible.  Anyway, this is what I mean by a parameter being less relevant- it might change the matching but there's nothing you can really _do_ with it.

A more relevant parameter would be something like:

```yaml
content:
  application/schema-instance+json; schema=https://example.com/schemas/foo/v1:
    schema: {$ref: "#/components/schemas/FooV1"}
  application/schema-instance+json; schema=https://example.com/schemas/foo/v2:
    schema: {$ref: "#/components/schemas/FooV2"}
```

Here it matters not because you need to decode (which I am reading as "parse", is this incorrect?) the media types any differently, but because the Media Type Objects have clearly different contents and therefore different behavior.

Does this make the utitlity of matching on parsed parameters more clear?

**Comment by karenetheridge** [comment]
> The more correct thing to do would be to parse the parameters and treat an entry with paramters as more specific than one without them. We very specifically note that we are comparing media ranges, which requires parsing them to determine that application/json is more specific than application/*. And, perhaps, application/foo+json is more specific than application/json

I was only describing issues with parameters appearing after the semicolon -- I wasn't speaking to the variation where a modifier is in the media-type itself. I don't handle that in any special way at all, although I agree that it should be possible to do so -- and indeed one might expect that a Content-Type of `application/schema+json` should be handled by a media-type entry of `application/json` if there is nothing more specific available.

> The spec says more specific takes precedence over more general

The spec only describes this in reference to wildcards -- e.g. `text/*` taking precedence over `text/plain`. It says nothing about modifiers appearing after a semicolon, or modifiers as in `application/schema+json`. (There are probably specific names for both of these in the RFCs, but I have not checked.)  I don't think it's reasonable that the reader is intended to assume that other variations of specific-over-generic should apply. If this was intended, it should enumerate the other ways in which specificity can be specified: namely the other two variants identified here in this thread.  Perhaps an OAD author might think this, but I object to the suggestion that this is something I ought to have thought of and provided for as an implementor, without specific mention of it.

Regarding charset -- of course it matters. `text/plain; charset=UTF-8` is very different from `text/plain; charset=ISO-8859-1` and characters will be different if they are not decoded correctly. I was saying that if we were taking the entire media-type string as a whole and defining different decoders for each one of these, then the code for these will be different, because the charset decoding needs to be taken into account as well as whatever other decoding is done for the media-type itself (which for text/plain is nothing, of course, but that's irrelevant to the example).  But if the parameters are considered separately, then they need to be supported outside the media-type decoder as well (and this is exactly what I do: after applying the media-type decoder to the content, I then apply charset decoding according to the charset parameter).

In summary: perhaps the matching should be done this way, but I don't think the spec allows for that today, and it either should do so (by being far more specific in its language), or we should use a different way to specify it.  For example if we add `parameters` under the media-type object I think that `Content-Type: foo/bar; modifier=foo` would _always_ match a media-type entry of `foo/bar`, and media-type names containing semicolons would never be valid -- that is, the parameter handling (if it is relevant to the API) MUST be done inside the media-type object for `foo/bar`.

As for handling things like `application/schema+json`, I don't know. I could see arguments either way for handling it either in a more specific media-type entry if it exists and falling back to `application/json` if it doesn't, or only matching `application/json` and using something in the media-type object; I'd want to consider more examples to think through the ramifications of going each way.

**Comment by handrews** [comment]
> The spec only describes this in reference to wildcards -- e.g. text/* taking precedence over text/plain. It says nothing about modifiers appearing after a semicolon

The spec describes it in terms of more or less specific media ranges, and gives _examples_ using wildcards.  But media ranges are defined by RFC9110 as follows:

```abnf
media-range = ( "*/*" / ( type "/*" ) / ( type "/" subtype ) ) parameters
parameter = parameter-name "=" parameter-value
parameter-name = token
parameter-value = ( token / quoted-string )
parameters = *( OWS ";" OWS [ parameter ] )
```

Since the examples using wildcards require parsing the string, and the general vs specific rule is about media ranges as described by RFC9110 (there's even a direct link), and that is a normative reference, then yes, I think it is clear that parameters are included in this.

Also, the whole point of media ranges is to use them in `Accept` to define was is acceptable, and the server needs to parse such things to determine what might be acceptable.  To do that, it needs to distinguish based on parameters, where ranges that have the same media type but different parameters might have different `q`-values.  So there is definitely an RFC9110 sense of comparing media ranges including parameters, and that cannot be done purely with string comparison.

> or modifiers as in `application/schema+json`

Yeah let's just exclude structured suffixes as, unlike parameters, they are not addressed by RFC9110.

> I was saying that if we were taking the entire media-type string as a whole and defining different decoders for each one of these, then the code for these will be different, because the charset decoding needs to be taken into account as well as whatever other decoding is done for the media-type itself (which for text/plain is nothing, of course, but that's irrelevant to the example). But if the parameters are considered separately, then they need to be supported outside the media-type decoder as well (and this is exactly what I do: after applying the media-type decoder to the content, I then apply charset decoding according to the charset parameter).

Decoding the content of the message in this sense is governed by the HTTP and media type specifications regarding `Content-Type`, not by the OAS/OAD.  Nothing in an OAD would change that, so unless you're implementing an HTTP content handler (lower-level than OpenAPI), I don't understand why you'd be doing anything here _as an OAS implementation_.  The only thing the OAS does is provide for matching `Content-Type` media _types_ to media _ranges_ that might not be exact matches.  Which is somewhat analogous to how `Accept` works.

**Comment by handrews** [comment]
I guess another way to look at this would be that given the following (note that I made the `schema` refs exactly match the parameters):

```yaml
content:
  text/plain: {}
  text/plain; charset=ISO-8859-1: {}}
  application/schema-instance+json; schema=https://example.com/schemas/foo/v1:
    schema: {$ref: https://example.com/schemas/foo/v1}
  application/schema-instance+json; schema=https://example.com/schemas/foo/v2:
    schema: {$ref: https://example.com/schemas/foo/v2}
```

`Content-Type` values of `text/plain`, `text/plain; charset=utf-8`, `text/plain; charset=ebcdic`, and `text/plain; charset=what-is-ths-who-knows` all MUST match `text/plain` in the `content` map.

Only a `Content-Type` value of `text/plain; charset=ISO-8859-1` (regardless of spaces involved after the semicolon) MUST match `text/plain; charset=ISO-8859-1`, and MUST NOT match `text/plain`.

But this parameter is used in the decoding process, prior to the text reaching the application layer.

On the other hand, a `Content-Type` of `application/schema-instance+json; schema=https://example.com/schemas/foo/v2` MUST match the corresponding entry, and the parameter has nothing to do with decoding (it's parsed as `application/json` and assumed UTF-8) but is relevant to the application layer.  The OAS automates some of that relevance by mapping that schema in a way that an OAS implementation can apply it correctly, without having to specifically understand the parameter.

But the application (above the OAS implementation) might well look at that parameter in `Content-Type` and make other decisions, for example because it is v2 of this schema, which might have implications beyond just picking which schema to use.

Assuming this is the complete `content` map, a `Content-Type` of `application/schema-instance+json; schema=https://example.com/other` MUST NOT match anything in my view.

**Comment by handrews** [comment]
I'm going to import this comment from @openreply-dleinhaeuser in #3162 so we can close that one as a dup of this:

-----

I have not been able to find any documentation that describes with sufficient detail how the Content-Type header of a request is matched against the entries of the Request Body Object.
In particular there is not a word lost on how to handle media ranges in the Request Body Object that have parameters, or if parameters are even supported there.
We had to learn that different OpenAPI validator implementations can have widely diverging behavior in this area.

Behaviors we have encountered include but are not limited to:

matching all specified parameters and their values with the parameters of the request content type
matching all specified parameters but not allowing extra parameters (many clients automatically add charset)
matching all specified parameters but also requiring the parameters to be specified in the same order as in the OpenAPI document (probably a bug)
matching all specified parameters but being strangely specific about whitespace (very likely a bug)
doing what type-is does (ignore parameters on request content type, break if there are parameters in the Request Body Object)
doing a simple string comparison between request Content-Type header and Request Body Object entry (definitely a bug)
This divergence prompted me to look up what the semantics are supposed to be, but I have been unable to find anything in the OpenAPI spec or in the commonly cited RFCs that provides clarity here.

Is there a definitive way this should be implemented? Is this unspecified on purpose?

I would be grateful if you could help me gain some clarity here.
